httpcore/_backends/curio.py

```diff
@@ -0,0 +1,203 @@
+import select
+from ssl import SSLContext, SSLSocket
+from typing import Optional
+
+import curio
+import curio.io
+
+from .._exceptions import (
+    ConnectError,
+    ConnectTimeout,
+    ReadError,
+    ReadTimeout,
+    WriteError,
+    WriteTimeout,
+    map_exceptions,
+)
+from .._types import TimeoutDict
+from .._utils import get_logger
+from .base import AsyncBackend, AsyncLock, AsyncSemaphore, AsyncSocketStream
+
+logger = get_logger("curio_backend")
+
+ONE_DAY_IN_SECONDS = 60 * 60 * 24
+
+
+def convert_timeout(value: Optional[float]) -> int:
+    return int(value) if value is not None else ONE_DAY_IN_SECONDS
+
+
+class Lock(AsyncLock):
+    def __init__(self) -> None:
+        self._lock = curio.Lock()
+
+    async def acquire(self) -> None:
+        await self._lock.acquire()
+
+    async def release(self) -> None:
+        await self._lock.release()
+
+
+class Semaphore(AsyncSemaphore):
+    def __init__(self, max_value: int, exc_class: type) -> None:
+        self.max_value = max_value
+        self.exc_class = exc_class
+
+    @property
+    def semaphore(self) -> curio.Semaphore:
+        if not hasattr(self, "_semaphore"):
+            self._semaphore = curio.Semaphore(value=self.max_value)
+        return self._semaphore
+
+    async def acquire(self, timeout: float = None) -> None:
+        exc_map: dict = {
+            curio.TaskTimeout: self.exc_class,
+        }
+        acquire_timeout: int = convert_timeout(timeout)
+
+        with map_exceptions(exc_map):
+            return await curio.timeout_after(acquire_timeout, self.semaphore.acquire())
```

Comment:
(Nit) Using a simpler `try/except` style here since there's only one exception mapping case, and this is also what we do in the `asyncio` backend.

Also once `convert_timeout()` is changed to return a `float`, it looks like we'll be able to drop the `int` type hint (which I think is not actually necessary even now?), and use the same variable name (since `timeout` will remain a `float`).

```suggestion
        timeout = convert_timeout(timeout)

        try:
            return await curio.timeout_after(timeout, self.semaphore.acquire())
        except curio.TaskTimeout:
            raise self.exc_class()
```

Reply:
Good point, going to change it